### PR TITLE
PIM-3704 : Only show variant groups sharing all their common attributes with selected products

### PR DESCRIPTION
**Pull request by grena**
| Q                    | A
| -------------------- | ---
| Bug fix?             | Y
| New feature?         | N
| BC breaks?           | N
| CI currently passes? | Y
| Tests pass?          | Y
| Scenarios pass?      | Y (build 1858)
| Checkstyle issues?*  | N
| PMD issues?**        | N
| Changelog updated?   | N
| Fixed tickets        | PIM-3704, PIM-3656, PIM-3739
| DB schema updated?   | N
| Migration script?    | N
| Doc PR               | N

This PR is a cloned PR of #2070,  with master as origin.

### CODE REVIEW

File: src/Pim/Bundle/CatalogBundle/Entity/Repository/GroupRepository.php [hunk 1]
```diff
@@ -107,49 +139,23 @@ public function getAllVariantGroupIds()
...
-        $qb = $this->createQueryBuilder('g');
+        $qb = $this->getAllVariantGroupsQB();
+        $variantGroups = $qb->innerJoin('g.attributes', 'attributes')
+            ->getQuery()->execute();
 
-        return $qb->innerJoin('g.type', 'type')
-            ->where($qb->expr()->eq('type.variant', ':variant'))
-            ->setParameter(':variant', true);
-    }
+        // @TODO: Try to put this block in the DQL query
```
Comment (sumbobyboys) [inline src/Pim/Bundle/CatalogBundle/Entity/Repository/GroupRepository.php]: nope, we should not merge @TODOs on master

File: src/Pim/Bundle/CatalogBundle/Repository/GroupRepositoryInterface.php [hunk 2]
```diff
@@ -70,9 +77,28 @@ public function getAllGroupsExceptVariant();
     public function getAllVariantGroups();
 
     /**
+     * Get all variant groups where $variantGroupIds are $include or not
+     *
+     * @param array $variantGroupIds
+     * @param bool  $include
```
Comment (jjanvier) [inline src/Pim/Bundle/CatalogBundle/Repository/GroupRepositoryInterface.php]: don't understand this parameter
bad naming of the parameter or of the method ?

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php [hunk 3]
```diff
@@ -155,36 +162,136 @@ public function getWarningMessages()
...
             ];
+        } elseif ($this->hasNoValidVariantGroupWarning()) {
+            $messages[] = [
+                'key'     => 'pim_enrich.mass_edit_action.add-to-variant-group.no_valid_variant_group',
+                'options' => []
+            ];
+        }
 
-            return $messages;
+        $invalidProducts = $this->getInvalidProductWarningInfos($this->skippedObjects);
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: good first move, but as this method is not public, you should pass $skipeddObjects as a variable (and rename it $skippedProducts)

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php [hunk 4]
```diff
@@ -155,36 +162,136 @@ public function getWarningMessages()
...
+        if ($invalidProducts) {
+            $messages[] = [
+                'key'     =>
+                    'pim_enrich.mass_edit_action.add-to-variant-group.already_in_variant_group_or_not_valid',
+                'options' =>
+                    ['%products%' => implode(', ', $invalidProducts)]
+            ];
         }
 
+        $skippedVariantGroups = $this->getSkippedVariantGroupWarningInfos($this->validVariantGroups);
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: same here :)

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php [hunk 5]
```diff
@@ -155,36 +162,136 @@ public function getWarningMessages()
...
+    {
+        return 0 === (int) $this->groupRepository->countVariantGroups();
+    }
+
+    /**
+     * Has a warning if there is no valid variant group.
+     *
+     * @return bool
+     */
+    protected function hasNoValidVariantGroupWarning()
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: Is think that `hasNoValidVariantGourp` will be a better name.

```
if ($this->hasNoVariantGroupWarning()) {
```
sounds a little bit odd

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php [hunk 6]
```diff
@@ -155,36 +162,136 @@ public function getWarningMessages()
...
+     * @param array $skippedProducts
+     *
+     * @return array
+     */
+    protected function getInvalidProductWarningInfos(array $skippedProducts)
+    {
         $invalidProducts = [];
-        foreach ($this->skippedObjects as $product) {
+
+        foreach ($skippedProducts as $product) {
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: nice

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php [hunk 7]
```diff
@@ -155,36 +162,136 @@ public function getWarningMessages()
...
+     *
+     * @param array $validVariantGroups
+     *
+     * @return array
+     */
+    protected function getSkippedVariantGroupWarningInfos(array $validVariantGroups)
+    {
+        $skippedVariantGroups = [];
+
+        // @TODO: Avoid getting all variant groups
```
Comment (sumbobyboys) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: We should avoid @TODOs

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php [hunk 8]
```diff
@@ -155,36 +162,136 @@ public function getWarningMessages()
...
+    }
+
+    /**
+     * Has a warning if there is no valid variant group.
+     *
+     * @return bool
+     */
+    protected function hasNoValidVariantGroupWarning()
+    {
+        return 0 === count($this->validVariantGroups);
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: seems it's not instanciated when the first call is performed

weird to do this kind of check on an instance variable

File: src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php [hunk 9]
```diff
@@ -155,36 +162,136 @@ public function getWarningMessages()
...
+        if ($products) {
+            $productIds = array_map(function ($product) {
+                return $product->getId();
+            }, $products);
+
+            $commonAttributeIds = $this->productMassActionRepo->findCommonAttributeIds($productIds);
+            $this->validVariantGroups = $this->groupRepository->getVariantGroupsByAttributeIds($commonAttributeIds);
+        }
+
+        return $this->validVariantGroups;
```
Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: nope

Comment (jjanvier) [inline src/Pim/Bundle/EnrichBundle/MassEditAction/Operation/AddToVariantGroup.php]: if you need to return something, drop your instance variable, it means you don't really need it

File: src/Pim/Bundle/CatalogBundle/Repository/GroupRepositoryInterface.php (hunk 2 above)
Comment (grena) [inline src/Pim/Bundle/CatalogBundle/Repository/GroupRepositoryInterface.php]: What do you think of `$whereIn` ?
`getVariantGroupsByIds(array $variantGroupIds, $whereIn = true);`
If set to false, it makes a `WHERE NOT IN` request. 

I know it's a tricky name, we talked about with @nidup

**Pull request merged**